### PR TITLE
[ZEPPELIN-5258] Add Knox realm group lookup to shiro auth service

### PR DESCRIPTION
**Pull request by Kimahriman**
### What is this PR for?
This PR aims to add support for Knox groups in ShiroAuthenticationService.getAssociatedRoles. This allows Knox realm groups (which uses hadoop groups under the hood) to be used for things like the Notebook API that doesn't use native Shiro group checks.

Additionally, I removed some unused parts of the Knox realm. I can undo this if desired, but I figured I would clean it up since it's not used.

### What type of PR is it?
Improvement

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-5258

### How should this be tested?
Added new UT and tested manually

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

File: conf/shiro.ini.template [hunk 1]
```diff
@@ -66,8 +66,6 @@ user3 = password4, role2
 #knoxJwtRealm.cookieName = hadoop-jwt
 #knoxJwtRealm.publicKeyPath = /etc/zeppelin/conf/knox-sso.pem
 #
-#knoxJwtRealm.groupPrincipalMapping = group.principal.mapping
-#knoxJwtRealm.principalMapping = principal.mapping
```
Comment (Kimahriman) [inline conf/shiro.ini.template:70]: In addition to being unused, this default config throws an error, as these are expected to be strings of the format

`<user1>=<mappeduser1>;<user2>=<mappeduser2>`

File: zeppelin-server/src/main/java/org/apache/zeppelin/realm/jwt/KnoxJwtRealm.java [hunk 2]
```diff
@@ -83,14 +78,6 @@
   @Override
   protected void onInit() {
     super.onInit();
-    if (principalMapping != null && !principalMapping.isEmpty()
-        || groupPrincipalMapping != null && !groupPrincipalMapping.isEmpty()) {
-      try {
-        mapper.loadMappingTable(principalMapping, groupPrincipalMapping);
```
Comment (Kimahriman) [inline zeppelin-server/src/main/java/org/apache/zeppelin/realm/jwt/KnoxJwtRealm.java:89]: `mapper` is never used anywhere else, so I removed everything related to it. Seems like something left over from copying some code from Knox.

File: zeppelin-server/src/main/java/org/apache/zeppelin/realm/jwt/KnoxJwtRealm.java [hunk 3]
```diff
@@ -241,7 +228,6 @@ protected AuthorizationInfo doGetAuthorizationInfo(PrincipalCollection principal
     /* return the groups as seen by Hadoop */
     Set<String> groups = null;
     try {
-      hadoopGroups.refresh();
```
Comment (Kimahriman) [inline zeppelin-server/src/main/java/org/apache/zeppelin/realm/jwt/KnoxJwtRealm.java:244]: Removed this so we could take advantage of hadoop's native group caching mechanisms that should be controllable via hadoop configurations

File: zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java [hunk 4]
```diff
@@ -138,8 +139,8 @@ private String extractPrincipal(Subject subject) {
   @Override
   public Collection<Realm> getRealmsList() {
     String key = ThreadContext.SECURITY_MANAGER_KEY;
-    DefaultWebSecurityManager defaultWebSecurityManager = (DefaultWebSecurityManager) ThreadContext.get(key);
-    return defaultWebSecurityManager.getRealms();
+    DefaultSecurityManager defaultSecurityManager = (DefaultSecurityManager) ThreadContext.get(key);
```
Comment (Kimahriman) [inline zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java:142]: Using a less specific class made it easier to test (hit a weird issue trying to create a DefaultWebSecurityManager and it seemed unnecessary)

File: zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java [hunk 5]
```diff
@@ -247,6 +248,9 @@ public boolean isAuthenticated() {
         } else if (ACTIVE_DIRECTORY_GROUP_REALM.equals(name)) {
           allRoles = ((ActiveDirectoryGroupRealm) realm).getListRoles();
           break;
+        } else if (realm instanceof KnoxJwtRealm) {
```
Comment (Kimahriman) [inline zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java:251]: Not sure if the other realms need to do string checks or should be changed to instanceof checks, but had to do it here for the test using the spy

**Comment by zjffdu** [comment]
Thanks for the contribution @Kimahriman Is doc here needs to be updated ? http://zeppelin.apache.org/docs/0.9.0/setup/security/shiro_authentication.html#knox-sso

File: zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java (hunk 5 above)
Comment (Reamer) [reply zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java:251]: You have an advantage when you do a string comparison.
There is a fixed string in the configuration and you can change the underlying implementation independently of the configuration.

**Comment by Kimahriman** [comment]
> Thanks for the contribution @Kimahriman Is doc here needs to be updated ? http://zeppelin.apache.org/docs/0.9.0/setup/security/shiro_authentication.html#knox-sso

Good catch, updated that doc as well

File: zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java (hunk 5 above)
Comment (Kimahriman) [reply zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java:251]: I don't really follow. We have to cast it to get the class specific behavior, so we're kinda relying on the implementation. And it prevents subclassing

Comment (Reamer) [reply zeppelin-server/src/main/java/org/apache/zeppelin/service/ShiroAuthenticationService.java:251]: Forget my previous stupid comment. You are right. I didn't look at all the lines of code.

**Comment by zjffdu** [comment]
Thanks @Kimahriman LGTM

**Comment by zjffdu** [comment]
Will merge if no more comment

**Pull request closed without merging**